Proxy address reuse and expiry

ylavic opened the pull request.

ylavic committed:
mod_proxy: Add/use ap_proxy_determine_address()

ylavic committed:
mod_proxy: Make ap_proxy_worker_name() pool-less and thread-safe.

ylavic committed:
mod_proxy: Factorize code in ap_proxy_get_worker_ex(), no functional change intended.

ylavic committed:
mod_proxy: Single "proxy:forward" worker and use of conf->pool to define/initialize workers.

ylavic committed:
mod_proxy: Update some comments, axe redundant socket_cleanup().

rpluem commented on modules/proxy/mod_proxy.h:
```diff
@@ -755,12 +765,13 @@ typedef __declspec(dllimport) const char *
  * @return         name
+ * @note Even though the returned name is non constant char*, the string
+ *       it points to is shared and should *not* be modified by the caller!
  */
```
How are the above changes related?

ylavic replied:
They are not :) As I said on dev@, I just happened to change them "by the way".
Only the first commit of this PR is really related actually.

rpluem commented on modules/proxy/mod_proxy_ftp.c:
```diff
@@ -1176,23 +1126,34 @@ static int proxy_ftp_handler(request_rec *r, proxy_worker *worker,
+    err = ap_proxy_determine_address("FTP", backend, connectname, connectport,
+                                     r, r->server);
+    if (APR_SUCCESS != err) {
+        proxy_ftp_cleanup(r, backend);
```
Why only `proxy_ftp_cleanup` and not `ftp_proxyerror` with  a  `DNS lookup failure for:` error message?

ylavic replied:
Argh, yes, either that or use NULL instead of r->server in the ap_proxy_determine_address() call to get an APLOG_ERR from there.
I'll go with the latter if my explanation below on why `(r && !s)` is not too twisted :)

ylavic replied:
Fixed in [52126bf](https://github.com/apache/httpd/pull/367/commits/52126bff6facd33ca10f03401565a09aa182018a), using ftp_proxyerror() finally.

rpluem commented on modules/proxy/mod_proxy.c:
```diff
@@ -227,6 +227,18 @@ static const char *set_worker_param(apr_pool_t *p,
+        }
+        worker->s->address_ttl = apr_time_from_sec(ttl);
+        worker->s->address_ttl_set = 1;
+    }
```
Why not using `ap_timeout_parameter_parse`?

ylavic replied:
This must be in seconds <= APR_INT32_MAX, but yes it could benefit from the "h" or "mi" suffixes still (but not "ms"..). I will use the below instead:
```
        /* Address TTL in seconds
         */
        apr_interval_time_t ttl;
        if (strcmp(val, "-1") == 0) {
            worker->s->address_ttl = -1;
        }
        else if (ap_timeout_parameter_parse(val, &ttl, "s") == APR_SUCCESS
                 && (ttl <= apr_time_from_sec(APR_INT32_MAX))
                 && (ttl % apr_time_from_sec(1)) == 0) {
            worker->s->address_ttl = ttl;
        }
        else {
            return "AddressTTL must be -1 or a number of seconds not "
                   "exceeding " APR_STRINGIFY(APR_INT32_MAX);
        }
        worker->s->address_ttl_set = 1;
```

ylavic replied:
I should add that with this PR, address_ttl=0 disables address and connection reuse completely for the worker, while an explicit address_ttl=-1|>0 (where -1 is infinite) allows for address reuse regardless of enable/disablereuse (which then relates to connection/socket reuse only).
IOW, this new parameter allows to distinguish betwwen enable/disablereuse and is_address_reusable, still when is_address_reusable is zero (either explicitely or implicitely for the generic workers) both the addresses and connections are never reused, same when enablereuse=0 and address_ttl= is not set, but an explicit address_ttl= to something not zero will reuse addresses and in this case enable/disablereuse (default or set) will apply to connections/sockets only.
I thought it was something desirable..

ylavic replied:
Fixed in [8ad4da6](https://github.com/apache/httpd/pull/367/commits/8ad4da692850d480bf2fd85518485596872e1b3b)

rpluem commented on modules/proxy/mod_proxy_balancer.c:
```diff
@@ -142,20 +142,18 @@ static int proxy_balancer_canon(request_rec *r, char *url)
-static void init_balancer_members(apr_pool_t *p, server_rec *s,
-                                 proxy_balancer *balancer)
+static void init_balancer_members(proxy_balancer *balancer,
+                                  server_rec *s, apr_pool_t *p)
```
Why changing the order of parameters?

ylavic replied:
Yeah, I changed this interface several times, passing the `proxy_server_conf*` etc to finally coming back to the initial parameters. Like this it aligns with ap_proxy_initialize_balancer/worker(), so I went with it..

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -1492,92 +1515,168 @@ static void socket_cleanup(proxy_conn_rec *conn)
-         * r->per_dir_config, r will likely get destroyed before this proxy
-         * conn is reused.
+        /* The current ssl section/dir config of the conn is not necessarily
+         * the one it will be reused for, so while the conn is in the reslit
```
```suggestion
         * the one it will be reused for, so while the conn is in the reslist
```

ylavic replied:
Thanks, will fix.

ylavic replied:
Fixed

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -1684,14 +1765,11 @@ static apr_status_t connection_destructor(void *resource, void *params,
-    return apr_pstrcat(p, "unix:", worker->s->uds_path, "|", worker->s->name, NULL);
+    (void)unused;
+    return worker->uds_name ? (char *)worker->uds_name : worker->s->name;
 }
```
Why the above change?

ylavic replied:
Well, it's not related, but not very wise either to race or leak on p here when we know the uds_name from the start.

I will not squash all the commits of this PR in the same svn commit (should this PR be fine) if that reassures you :)

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -3544,12 +3823,10 @@ static int proxy_connection_create(const char *proxy_function,
                      "new connection to %pI (%s)", proxy_function,
                      backend_addr, conn->hostname);
-        /* XXX: Will be closed when proxy_conn is closed */
-        socket_cleanup(conn);
```
Why isn't this needed any longer?

ylavic replied:
Every caller of `ap_proxy_connection_create[_ex]()` should release/invalidate the connection on failure already, like with any other ap_proxy_ function taking a proxy_conn_rec which fails? We do this upstream already afaict.

rpluem replied:
Good point, but in order to be save shouldn't we do at least a `conn->close = 1` here in case the caller only releases the connection but forgets to update the `close` flag before? But I agree that the current consumers in our code do it properly.

ylavic replied:
Added [97ce1e9](https://github.com/apache/httpd/pull/367/commits/97ce1e996b29d8666ae6778ebe4e1feff0986650) to be consistent on how we close the socket on reuse failure.

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -2620,6 +2697,248 @@ PROXY_DECLARE(int) ap_proxy_release_connection(const char *proxy_function,
+                if (rv != APR_SUCCESS) {
+                    PROXY_THREAD_UNLOCK(worker);
+                    apr_pool_destroy(pool);
+                    if (r && !s) {
```
Is this the healthcheck worker case?

ylavic replied:
It's for when the user wants an ERR log (and some error details in r->notes) generated by ap_proxy_dermine_address() directly. In this PR, it applies to the calls in ap_proxy_determine_connection() and ftp (well I missed that one actually as you noticed above), while the calls in hcheck and ap_proxy_connect_backend() have no `r` so use `s` only.
So `(r && !s)` calls ap_proxyerror() (thus APLOG_ERR and notes), while `r` or `s` alone issue an APLOG_DEBUG only. Yeah I know it's not a pretty interface, but I didn't want/feel to use an explicit parameter for this, and it's in the dox! :p

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -2620,6 +2697,248 @@ PROXY_DECLARE(int) ap_proxy_release_connection(const char *proxy_function,
+            PROXY_THREAD_LOCK(worker);
+
+            /* Re-check under the mutex, things may have changed. */
+            address = worker->address;
```
Why don't we need to use `AP_VOLATILIZE_T` here?

ylavic replied:
It'd probably be consistent/clearer to do so, though worker->address can only change under the mutex and I don't think that the compiler is allowed to reorder this instruction much here since it's preceded by a barrier (mutex lock) and followed by a control/address dependency (the `&address->expiry`).

rpluem replied:
My worry is that the compiler stored worker->address in a register when doing the assignment in line 2791 and reuse this value from the register after the lock. But this worry might be unfounded.

ylavic replied:
I don't think that the compiler is allowed to do that because it'd move an instruction from inside to outside the critical section, which is quite against the principle of critical sections. The compiler can move instructions from outside to inside, but not here either for `address` because it's used in the condition to access the CS (data and address dependency) and it's volatilized anyway to keep the instruction just before the condition.
But to be clear/consistent I'll change to use volatile access inside the CS too.

ylavic replied:
Done in [8ad4da6](https://github.com/apache/httpd/pull/367/commits/8ad4da692850d480bf2fd85518485596872e1b3b)

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -2620,6 +2697,248 @@ PROXY_DECLARE(int) ap_proxy_release_connection(const char *proxy_function,
+
+            /* Update worker->cp->addr if it's not set (or reset) */
+            if (!AP_VOLATILIZE_T(apr_sockaddr_t *, worker->cp->addr)) {
+                worker->cp->addr = address->addr;
```
How do we ensure it never expires even if it got reset in the meantime?

ylavic replied:
The first address/DNS lookup will never expire because it has a refcount of 2, but if some (third party) module plays by resetting worker->cp->addr arbitrarily after the startup it will be set to the current address, which will expire when a new one is needed and the worker or last connection using it releases it.
I don't think it's ever been safe to reset worker->cp->addr, so we should be good?

rpluem replied:
I agree that using setting `worker->cp->addr` to `NULL` as an indicator was not a good idea and given that this possibility is probably only available in one patch release of 2.4. I am personally happy to drop this. I only feel unsure if we could do it for formal reasons. But probably as long as no one complains we get away with it :-p.

ylavic replied:
I don't see where we do this in 2.4, it's trunk changes only for now and addressed by this PR so we should be good ;)

rpluem replied:
My bad. Missed that it was never part of 2.4.x.

ylavic committed:
Address Ruediger's comments (to squash in "Add/use ap_proxy_determine_address()")

ylavic committed:
Address Ruediger's comment (to squash in "Update some comments, ...")

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -403,8 +419,12 @@ PROXY_DECLARE(char *)
 }
 
-PROXY_DECLARE(int) ap_proxyerror(request_rec *r, int statuscode, const char *message)
+static int ap_proxyerror_ex(request_rec *r, int statuscode, const char *message,
```
Is this the correct name for a `static` function? I think it should be either `proxyerror_core` (at least dropping the `ap_` prefix, the further part of the name I don't care that much) or we should make it part of the API and properly declare it via `PROXY_DECLARE` as non static function and add it to `mod_proxy.h`.

ylavic replied:
Done in [8ad4da6](https://github.com/apache/httpd/pull/367/commits/8ad4da692850d480bf2fd85518485596872e1b3b)

ylavic committed:
mod_proxy: Consistently close the socket on failure to reuse the connection.

ylavic committed:
Address Ruediger's comment (to squash in "Add/use ap_proxy_determine_address()")

ylavic committed:
Merge branch 'trunk' of https://github.com/apache/httpd into proxy_address_reuse_and_expiry

ylavic committed:
Merge branch 'trunk' of https://github.com/apache/httpd into proxy_address_reuse_and_expiry

ylavic committed:
mod_proxy: Set new APLOGNOs (to squash in "Add/use ap_proxy_determine_address()")

rpluem commented:
Let's go for it.

rpluem commented:
@ylavic : Do you care to commit to trunk?

rpluem commented:
@ylavic : Do you care to commit to trunk?

ylavic committed:
address_ttl in seconds.

ylavic committed:
Fix some atomic/lifetime issues.

ylavic committed:
Add flags to ap_proxy_determine_address() for better extendability.

ylavic committed:
Add ap_proxy_worker_get_name() deprecating ap_proxy_worker_name()

ylavic committed:
Minor bump for ap_proxy_determine_address()

ylavic commented:
> Do you care to commit to trunk?

Yeah, sorry about the delay (quite busy times..).
I just pushed some more commits for what I could find (re-)reading the changes a few months later, the plan is to squash the many commits in bigger related ones with better commit messages (I'll do this locally) and then push that to trunk. Hopefully tonight..

ylavic committed:
Add PROXY_DETERMINE_ADDRESS_CHECK flag for ap_proxy_determine_address()
to check if the address changed only, and if not return APR_EEXIST.
If it changed update the worker address as usual for every one to
know/update.

This is used by ap_proxy_connect_backend() to check if it should try
with some new address(es) on failure to connect() first with the current
one(s), and avoid expiring the worker address globally/spuriously for
connect() errors potentially not related to expired address(es).

ylavic commented:
The last commit ([176de10](https://github.com/apache/httpd/pull/367/commits/176de10e2bd63b7d3d025010a3f8f7072f487a29)) is to avoid expiring the current backend address forcibly and maybe spuriously if ap_proxy_connect_backend() fails on the existing address(es) for something else than some address(es) change. WDYT?

ylavic committed:
Fix proxy_addrs_equal().

darmbrust commented:
Is there an easy way to figure out what version this patch landed in?  I see the docs for it are already advertised, but the docs don't say when it became a valid option.  I don't see it mentioned in the changelog.

Also, if I'm understanding this feature correctly, it will be a better solution to a problem folks have where load balancers rotate IPs:
https://serverfault.com/questions/870026/apache-honouring-the-dns-ttl-in-proxy-pass
The current suggestion is to disablereuse=On

Does this section of the workers doc need to be updated with this feature?
https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#workers

Specifically:

**
DNS resolution for origin domains

DNS resolution happens when the socket to the origin domain is created for the first time. When connection reuse is enabled, each backend domain is resolved only once per child process, and cached for all further connections until the child is recycled. This information should to be considered while planning DNS maintenance tasks involving backend domains. Please also check ProxyPass parameters for more details about connection reuse. 
**

How does the above interact with this new feature?  Will it remove connections when at age address_ttl?  Or, will it take a failure on a connection, and then resolve dns if address_ttl is expired before creating a new connection?

rpluem commented:
> Is there an easy way to figure out what version this patch landed in? I see the docs for it are already advertised, but the docs don't say when it became a valid option. I don't see it mentioned in the changelog.

It will be part of the next 2.4.x release.

> 
> Also, if I'm understanding this feature correctly, it will be a better solution to a problem folks have where load balancers rotate IPs: https://serverfault.com/questions/870026/apache-honouring-the-dns-ttl-in-proxy-pass The current suggestion is to disablereuse=On
> 
> Does this section of the workers doc need to be updated with this feature? https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#workers
> 
> Specifically:
> 
> ** DNS resolution for origin domains
> 
> DNS resolution happens when the socket to the origin domain is created for the first time. When connection reuse is enabled, each backend domain is resolved only once per child process, and cached for all further connections until the child is recycled. This information should to be considered while planning DNS maintenance tasks involving backend domains. Please also check ProxyPass parameters for more details about connection reuse. **
> 

This would need some update.

> How does the above interact with this new feature? Will it remove connections when at age address_ttl? Or, will it take a failure on a connection, and then resolve dns if address_ttl is expired before creating a new connection?

It will close connections when the TTL of the DNS entry is expired or if the the address changed in the meantime as another connection triggered a new lookup which resulted in a different address and recreate them.
